### PR TITLE
Update Smarty addTemplateDir function signature to future smarty

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------

The `addTemplateDir` function is not in Smartyv2. However, it is added in Smarty v3 with the signature below (which is the same in v4), This updates our functions signature to the signature that will someday be on the parent class, calling the parent class if it is there.

```
/**
     * Add template directory(s)
     *
     * @param string|array $template_dir directory(s) of template sources
     * @param string       $key          of the array element to assign the template dir to
     * @param bool         $isConfig     true for config_dir
     *
     * @return Smarty          current Smarty instance for chaining
     */
    public function addTemplateDir($template_dir, $key = null, $isConfig = false)
   ``` 

Before
----------------------------------------
Function signature does not match future parent

After
----------------------------------------
Now it does

Technical Details
----------------------------------------
We may never upgrade smarty ... but this seems like sensible future proofing

Comments
----------------------------------------

### CODE REVIEW

**Comment by totten** [comment]
TLDR: Consider an earlier `return`. Add more warnings to docblocks.

(1) If `parent::addTemplateDir()` updates  `$this->template_dir`, then the subsequent steps (`array_unshift($this->template_dir)`) would create duplicate registrations, no? Seems like it should be either/xor. Maybe you can return at that point?

(2) The `array_unshift()` prepends a directory. The `parent::addTemplateDir()` appends a directory. Probably works the same for 90%+ of extensions/configurations; I suppose the impacted edge-case would be for overridden `*.tpl`s. That's not really a fixable conflict, and I won't shed any tears about overidden `*.tpl`s. But a disclaimer is warranted. Maybe like:

> When using addTemplateDir(), the order of directories is not guaranteed, and it may change in future versions/configurations. TPL overrides will not be reliable.

(3) This implementation passes-through `$key` and `$isConfig` on Smarty v3/v4 -- but ignores them on Smarty v2. From a consumer POV, this makes the contract flaky. (*"The contract said I could add config-dir items, but it wrongly puts them in the template-dir path!"*) I'd suggest putting a disclaimer on the docblock (e.g. describe the new params as `EXPERIMENTAL` or some such).

**Comment by eileenmcnaughton** [comment]
@totten I agree re 1 & fixed it.

For 3 I added (smarty 3+) to the docblock comment

I didn't do anything for 2 because that logic is unchanged by this PR & I didn't want to bring it into scope (the variable name is changed in that line but not the logic)

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
Abandoning this as I added to https://github.com/civicrm/civicrm-core/pull/25362 - since I think unstalling one PR might be easier than unstalling 2

**Comment by eileenmcnaughton** [comment]
Re-opened as the other failed tests & I addressed @totten's feedback 'last year' (thanks github) - I thought it might be easier to merge this in  - but the other one has some challenges so let's get this merged & deal with getting the smarty mixin to the point of being useful separately

**Comment by eileenmcnaughton** [comment]
@totten I feel like this was agreed but it still just needs merging. There is the follow on work which still needs more work but it should not affect this part

**Comment by seamuslee001** [comment]
ok this seems fine to me

**Comment by eileenmcnaughton** [comment]
thanks @seamuslee001
